File: app/app/src/main/java/com/dairydash/utils/PricingEngine.kt
package com.dairydash.utils

/**
 * Kotlin port of pricing_engine.py
 * Calculates milk price per litre based on state, fat%, SNF%, quantity, and quality score.
 */
object PricingEngine {

    data class PricingResult(
        val success: Boolean,
        val state: String = "",
        val basePricePerLitre: Double = 0.0,
        val finalPricePerLitre: Double = 0.0,
        val qualityScore: Double = 0.0,
        val quantity: Double = 0.0,
        val totalPayment: Double = 0.0,
        val error: String = ""
    )

    fun calculatePrice(
        state: String,
        fat: Double,
        snf: Double,
        quantity: Double,
        qualityScore: Double
    ): PricingResult {
        return try {
            val basePricePerLitre = when (state) {
                "Tamil Nadu"  -> (fat * 7.0) + (snf * 3.0)
                "Kerala"      -> (fat * 6.8) + (snf * 3.2)
                "Karnataka"   -> {
                    val basePrice = 30.0
                    val fatIncentive = if (fat > 3.5) (fat - 3.5) * 2.0 else 0.0
                    val subsidy = 5.0
                    basePrice + fatIncentive + subsidy
                }
                "Gujarat"     -> (fat * 6.5) + (snf * 4.0)
                "Maharashtra" -> fat * 8.0
                "Punjab"      -> (fat * 7.5) + (snf * 2.5)
                "Haryana"     -> (fat * 7.2) + (snf * 2.8)
                else          -> (fat * 6.5) + (snf * 3.0) // Default fallback
            }

            // Apply quality score impact: score 100 → full price; score 50 → 50% of price
            val finalPricePerLitre = basePricePerLitre * (qualityScore / 100.0)
            val totalPayment = finalPricePerLitre * quantity

            PricingResult(
                success = true,
                state = state,
                basePricePerLitre = round2(basePricePerLitre),
                finalPricePerLitre = round2(finalPricePerLitre),
                qualityScore = qualityScore,
                quantity = quantity,
                totalPayment = round2(totalPayment)
            )
        } catch (e: Exception) {
            PricingResult(success = false, error = e.message ?: "Unknown error")
        }
    }

    fun getFormulaText(state: String): String = when (state) {
        "Tamil Nadu"  -> "Price = (Fat × 7) + (SNF × 3)"
        "Kerala"      -> "Price = (Fat × 6.8) + (SNF × 3.2)"
        "Gujarat"     -> "Price = (Fat × 6.5) + (SNF × 4)"
        "Maharashtra" -> "Price = Fat × 8"
        "Karnataka"   -> "Price = Base(30) + Fat Incentive(>3.5) + Subsidy(5)"
        "Punjab"      -> "Price = (Fat × 7.5) + (SNF × 2.5)"
        "Haryana"     -> "Price = (Fat × 7.2) + (SNF × 2.8)"
        else          -> "Standard Formula"
    }

    private fun round2(value: Double): Double =
        Math.round(value * 100.0) / 100.0
}

File: pricing_engine.py
import sys
import json

def calculate_price(data):
    try:
        state = data.get('state', '').strip()
        fat = float(data.get('fat', 0))
        snf = float(data.get('snf', 0))
        quantity = float(data.get('quantity', 0))
        quality_score = float(data.get('quality_score', 100))

        base_price_per_litre = 0

        # State-wise formulas
        if state == "Tamil Nadu":
            base_price_per_litre = (fat * 7) + (snf * 3)
        elif state == "Kerala":
            base_price_per_litre = (fat * 6.8) + (snf * 3.2)
        elif state == "Karnataka":
            base_price = 30 # Base price assumption
            fat_incentive = (fat - 3.5) * 2 if fat > 3.5 else 0
            subsidy = 5
            base_price_per_litre = base_price + fat_incentive + subsidy
        elif state == "Gujarat":
            base_price_per_litre = (fat * 6.5) + (snf * 4)
        elif state == "Maharashtra":
            base_price_per_litre = fat * 8
        elif state == "Punjab":
            base_price_per_litre = (fat * 7.5) + (snf * 2.5)
        elif state == "Haryana":
            base_price_per_litre = (fat * 7.2) + (snf * 2.8)
        else:
            # Default fallback formula
            base_price_per_litre = (fat * 6.5) + (snf * 3)

        # Apply Quality Score impact
        # If score is 100, they get 100% of the price.
        # If score is 50, they get 50% of the price.
        final_price_per_litre = base_price_per_litre * (quality_score / 100.0)
        
        total_payment = final_price_per_litre * quantity

        result = {
            "success": True,
            "state": state,
            "base_price_per_litre": round(base_price_per_litre, 2),
            "final_price_per_litre": round(final_price_per_litre, 2),
            "quality_score": quality_score,
            "quantity": quantity,
            "total_payment": round(total_payment, 2)
        }
        return result

    except Exception as e:
        return {"success": False, "error": str(e)}

if __name__ == "__main__":
    # Read JSON input from standard input
    try:
        input_data = sys.stdin.read()
        data = json.loads(input_data)
        result = calculate_price(data)
        print(json.dumps(result))
    except Exception as e:
        print(json.dumps({"success": False, "error": str(e)}))
